Support more than one request in report upload API (#686)
ietf-wg-ppm/draft-ietf-ppm-dap: draft-ietf-ppm-dap.md

struct {} Empty;
~~~

Errors that occurred during processing individual reports are represented
by the following enum:

~~~ tls-presentation
enum {
  reserved(0),
  batch_collected(1),
  report_replayed(2),
  report_dropped(3),
  hpke_unknown_config_id(4),
  hpke_decrypt_error(5),
  vdaf_prep_error(6),
  task_expired(7),
  invalid_message(8),
  report_too_early(9),
  task_not_started(10),
  outdated_config(11),
  (255)
} ReportError;
~~~

> 

Reviewer: ```suggestion
```

  HpkeCiphertext leader_encrypted_input_share;
  HpkeCiphertext helper_encrypted_input_share;
} Report;

struct {

>   Report reports<0..2^32-1>;

Reviewer: My understanding was that one of the motivations here is to deduplicate public extensions across reports, so I would think we'd have a field for that here.

Author: I was a bit worried that it may violate the assumption of upload API to support (1..N) reports. Requests to upload more than one report will have extra `public_extensions` field. Or, if we decide to always include them, requests to upload 1 report will need to omit `public_extensions` or duplicate them across two fields.

Reviewer: I think we should hoist `public_extensions` out of `struct ReportMetadata` and up into `struct UploadRequest` (though of course we have to pay careful attention to how per-report AAD is computed to make sure all extensions are still mixed in). Suppose reports with no extensions get uploaded. Right now we have:

```
struct {
 ReportID report_id; // 16 bytes
 Time time; // 8 bytes
 Extension public_extensions<0..2^16-1>; // 2 bytes for empty
} ReportMetadata;

struct {
 ReportMetadata report_metadata; // 26 bytes
 opaque public_share<0..2^32-1>; 
 HpkeCiphertext leader_encrypted_input_share;
 HpkeCiphertext helper_encrypted_input_share;
} Report;

struct {
  Report reports<0..2^32-1>; // 26 bytes + public share and encrypted shares per report
} UploadRequest;
```

If we hoist the public extensions out, we get (`struct Report` doesn't change):

```
struct {
    ReportID report_id; // 16 bytes
    Time time; // 8 bytes
} ReportMetadata;

struct {
    Extension public_extensions<0..2^16-1>; // 2 for empty
    Report reports<0..2^32-1>; // 24 bytes + public share and encrypted shares per report
}
```

So it seems to me we save 2 bytes per report past the first one by hoisting out public extensions.

IMO de-duplicating the repeated public extensions is the principal reason to make this change. If we don't do that, what's the motivation here?

Reviewer 2: If the public extensions really are repeated, HTTP compression should handle them fine, as long as the report structures are all aligned the same way anyway. Seems like that should be testable, and easier than this?

Reviewer: Request body compression is a really good point I hadn't considered. I think it makes more sense to rely on that than to allow for extensions in yet another place.

Reviewer 3: I like the basic idea @cjpatton described for compression here  (I would call it `common_extensions` rather than `public_extensions`), but do the reports really need to have a length prefix?  That could be a compromise between different approaches.  It's not like we can't rely on the outer message framing (like Content-Length) as an indication of the end of the message.  It saves 4 bytes.  And it opens up the possibility of indeterminate length requests, if an implementation is able to handle that.

Reviewer 4: Removing the length prefix sounds good to me, but I don't know how to spell this in TLS' presentation language.@martinthomson do you have a suggestion?

We could just say in prose: "the payload of the HTTP request is a sequence of `Report` structures concatenated together". This is a little ugly but I suppose it would work.

Reviewer 3: From memory:

```
~~~ tls-syntax
struct Whatever {
  HeaderType header;
  ValueType values[value_count];
};
~~~
```

That is, you use the fixed-size vector syntax, but use a variable instead of a number.

> The size of the vector may be specified at documentation time or left unspecified until runtime.  

-- https://datatracker.ietf.org/doc/html/rfc8446#section-3.4

Of course, this isn't perfectly clear there, because it doesn't say how to do that.

Reviewer 4: That works for me!

Author: I've changed <0...2^32-1> to [2^32-1] as it seems the most appropriate syntax for this

> Here, T' occupies n bytes in the data stream, where n is a multiple
   of the size of T.

Reviewer 3: > I've changed <0...2^32-1> to [2^32-1]

The brackets indicate a fixed size, so that says that you will always have 4GiB of data. That's probably not what you intended.  You need a variable, like: `Whatever thing[VARIABLE_NAME_HERE];`

Author: changed this to [report_count]/[status_count]

### Upload Request

Clients upload reports by sending a POST to `{leader}/tasks/{task-id}/reports`.

> The body is a `UploadRequest`, with media type "application/dap-upload-req", structured as

Reviewer: ```suggestion
The body is an `UploadRequest`, with media type "application/dap-upload-req", structured as
```

  HpkeCiphertext leader_encrypted_input_share;
  HpkeCiphertext helper_encrypted_input_share;
} Report;

struct {

>   Report reports[report_count];

Reviewer: `report_count` is misleading. Per RFC 8446 vector syntax ([1](https://datatracker.ietf.org/doc/html/rfc8446#section-3.4)), the value in square brackets is the length *in bytes*, not the count of elements in the vector. Let's use `content_length` as in https://github.com/ietf-wg-ppm/draft-ietf-ppm-dap/commit/4867378c805dfb961aacd4be0c975d26946c1922.

* `helper_encrypted_input_share` is the Helper's encrypted input share.

The length of the sequence is specified at runtime, but it must not exceed

> 2^32 - 1 bytes.

Reviewer 4: I don't think the 2GB limit we impose elsewhere in the draft applies here. The limit exists only because we need to specify the length of the length prefix of fields in the wire format of messages. since we're not encoding the length prefix, we don't need to specify an upper bound ... at least not for the same reason.

> 2^32 - 1 bytes.

Reviewer 4: I think we should just remove this paragraph.

> 2^32 - 1 bytes.

Reviewer: No, I think this is a misinterpretation of RFC 8446 vectors. https://datatracker.ietf.org/doc/html/rfc8446#section-3.4:

```
T T'[n];

   Here, T' occupies n bytes in the data stream, where n is a multiple
   of the size of T.
```

> 2^32 - 1 bytes.

Reviewer 3: I'm not happy with mandating the use of Content-Length.  We discourage it in HTTP/2 and HTTP/3 and it would discourage the use of chunked transfer encoding in HTTP/1.1, which is a perfectly good thing to use.

Yes, it deprives servers of certainty.  They need to learn to get over that.

> 2^32 - 1 bytes.

Reviewer: @martinthomson OK, I'm willing to take your word on this being a problem in HTTP/2 or 3. If there's no Content-Length and no length in the protocol-level message, how then are servers supposed to know when the message ends?

> 2^32 - 1 bytes.

Reviewer: Oh, I think I get what you're saying: a server _could_ use `Content-Length` if it happens to be using HTTP 1.1 without chunked transfer, but we don't want to force all possible configurations to do so. So I guess we'd have to write some text here saying roughly "the length of the vector has to be determined from the transport layer, which could be Content-Length in some HTTP/1.1 settings, END_STREAM in HTTP/2 or something else"?

> 2^32 - 1 bytes.

Reviewer 3: Both HTTP/2 and HTTP/3 have a clear end of stream indication, yes.  The [latest text](https://datatracker.ietf.org/doc/html/rfc9113#section-8.1.1-2) in RFC 9113 doesn't exactly forbid the use of `Content-Length`, but it does imply strongly that you can and should avoid it.  At a minimum, it's redundant.  At worst, it disagrees with the other parts of the protocol, causing errors.

> a server could use Content-Length if it happens to be using HTTP 1.1 without chunked transfer, but we don't want to force all possible configurations to do so

Precisely.  What you said is where I'd look to go.

} InputShareAad;
~~~

~~~ tls-presentation
struct {

> ReportID id;

Reviewer 5: Would it be helpful in terms of efficiency to omit the ID here?  It can be inferred from the order of responses in the vector.

Reviewer: Based on the "Failed upload of 1/2 reports submitted in one bulk upload" example, it seems like the intent might be that the response can include `ReportUploadStatus` values only for the reports that weren't successfully uploaded, and at that point you do need to report ID to be able to tell which it was. I kinda like that, because as we expect the vast majority of reports to be uploaded without issue, it massively cuts down on the size of responses. 

You might be able to save even more room on the wire by including an index into the vector in the `UploadRequest` instead of the report ID, but that would require clients (or that entity which uploads, which may not be a client in proxy arrangements) to maintain an index of upload requests to report IDs, which might not be tractable.

> ReportID id;

Reviewer 4: Need a tab here (4 spaces). Likewise below.
```suggestion
    ReportID id;
```

Author: it seems that we use [2 spaces](https://github.com/ietf-wg-ppm/draft-ietf-ppm-dap/blob/main/draft-ietf-ppm-dap.md?plain=1#L1100C1-L1100C3) everywhere? 

still need to fix this though

ReportError error;
} ReportUploadStatus;

struct {
ReportUploadStatus status<0..2^32-1>;
} UploadResponse;
~~~

If the upload request is malformed, the Leader aborts with error
`invalidMessage`.

If the Leader does not recognize the task ID, then it aborts with error
`unrecognizedTask`.

> Otherwise, the Leader responds with HTTP status 200, a body consisting
> of `UploadResponse` and the media type `application/dap-report-resp`. If

Reviewer: Per RFC 9205, we don't prescribe specific HTTP statuses, but leave it up to implementations to choose which success status to use.
```suggestion
Otherwise, the Leader responds with a body consisting
of an `UploadResponse` and the media type `application/dap-report-resp`. If
```

the entire upload has been successfully completed, the Leader MAY respond with

> an empty body.

Reviewer 4: I'm not in favor of this MAY. It's better I think if there's only one way to handle the response: expect an `UploadResponse`.

Reviewer: I don't agree. We use empty response bodies in the aggregation and collection interactions to indicate that aggregation, aggregate share and collection jobs are being handled asynchronously, so using an empty body here seems fine, and saves a few bytes on the wire. The encoding of an empty upload response is bigger than nothing.

Reviewer 4: Well, there's an important difference I think.

During aggregation, an empty response indicates async handling, which means the Leader needs to poll later. But there is no async processing of the upload request. Should there be?

Reviewer: Sure, it's not exactly the same, but the point is that we already assign semantic meaning to the absence of a body. Further, the real issue here is that there's no point in constructing and parsing a message that conveys no information (the message would just be 4 0 values bytes indicating an empty vector of reports) when literally nothing would serve just as well.

Reviewer 4: What I think is odd about this is that we have to special case the empty body. That is the code is something like:

```rust
fn parse_request_body(bytes: &[u8]) -> Vec<ReportStatus> {
   if bytes.empty() {
      return Vec::new();
   }
   parse_report_status_list(bytes)
}
```

But tbh I don't feel very strongly about this. I'm ok with whatever we we do here.

For each report in the upload, the Leader creates `ReportUploadStatus` struct
and adds the `ReportId` to it. If there was no error during the processing,

> the Leader sets the `error` to `none`.

Reviewer 4: Thanks for implementing this suggestion here @akoshelev. However, after thinking this more, I think this is may not a good idea. The problem I see is that during the aggregation flow, a `ReportError error = none` is not a valid response. If there was really no report error, then the aggregator would expect a different response than a `ReportError`.

That said, I think you and @tgeoghegan have already suggested a much better idea (https://github.com/ietf-wg-ppm/draft-ietf-ppm-dap/pull/686/files#r2070551993): if a report was uploaded successfully, then simply don't include it in the `UploadResponse`.

} InputShareAad;
~~~

~~~ tls-presentation
struct {
  ReportID id;
  ReportError error;
} ReportUploadStatus;

struct {

>   ReportUploadStatus status<0..2^32-1>;

Reviewer 3: Same comment as above.  There's no need for another length on this.

If the Leader does not recognize the task ID, then it aborts with error
`unrecognizedTask`.

Otherwise, the Leader responds with a body consisting of an `UploadResponse` and
the media type `application/dap-upload-resp`. The structure of the response
is shown below:

~~~ tls-presentation
struct {
  ReportID id;
  ReportError error;
} ReportUploadStatus;

struct {

>   ReportUploadStatus status[status_count];

Reviewer: As above, `status_count` is misleading because the value here is a number of bytes. Let's use `content_length`. And we should make it it clear that the response's `Content-Length` header MUST match the value used here.

} UploadResponse;
~~~

The Leader only includes reports that failed processing in the response.
Reports that are accepted do not have a response.

> Reports in the response follow the same order as specified in the request.

Reviewer: ```suggestion
Reports in the response MUST appear in the same order as in the request.
```

For each report that failed to upload, the Leader creates a
`ReportUploadStatus` struct and includes the `ReportId` from the input and a

> `ReportError` {{basic-definitions}} that describes the failure. The length of

Reviewer: ```suggestion
`ReportError` ({{basic-definitions}}) that describes the failure. The length of
```

this sequence is always less than or equal to the length of the upload sequence.

If the Leader does not recognize the `config_id` in the encrypted input share,

> it sets the error field to `outdated_config`
> (see {{aggregation-helper-init}}). When the Client receives an
> `outdated_config` error, it SHOULD invalidate any cached `HpkeConfigList` and

Reviewer: ```suggestion
it sets the error field to `outdated_config`. When the Client receives an
`outdated_config` error, it SHOULD invalidate any cached `HpkeConfigList` and
```
Is this reference correct? The `ReportError` enum was moved up in the document. Actually, since the preceding paragraph already points to {{basic-definitions}}, let's delete this altogether.

retry with a freshly generated `Report`. If this retried upload does not
succeed, the Client SHOULD abort and discontinue retrying.

If a report's ID matches that of a previously uploaded report, the Leader MUST
ignore it. In addition, it MAY alert the Client with setting the
error field to `report_replayed`.

The Leader MUST ignore any report pertaining to a batch that has already been
collected (see {{replay-protection}} for details). The Leader MAY also return
`report_replayed` error for this report.

The Leader MUST ignore any report whose timestamp is outside of the task's
`time_interval`. When it does so, it SHOULD also set the error field to
`report_dropped` for the corresponding report in the response.

The Leader may need to buffer reports while waiting to aggregate them (e.g.,
while waiting for an aggregation parameter from the Collector; see
{{collect-flow}}). The Leader SHOULD NOT accept reports whose timestamps are too
far in the future. Implementors MAY provide for some small leeway, usually no
more than a few minutes, to account for clock skew. If the Leader rejects a
report for this reason, it SHOULD indicate it with `report_too_early` error.

> In this situation, the Client MAY re-upload the report later on.

Reviewer: Let's use consistent language for error handling to avoid implying that the server should do different things in different cases. "set the error field to `XXX` for the corresponding report in the response" should work

#### Example

Successful upload

~~~ http
POST /leader/tasks/8BY0RzZMzxvA46_8ymhzycOB9krN-QIGYvg_RsByGec/reports
Host: example.com
Content-Type: application/dap-report-req

encoded(struct {
  reports = [struct {
    report_metadata = struct {
      report_id = [0x0a, 0x0b, 0x0c, 0x0d, ...],
      time = 1741986088,
      public_extensions = [0x00, 0x00],
    } ReportMetadata,
    public_share = [0x0a, 0x0b, ...],
    leader_encrypted_input-share = struct {
      config_id = 1,
      enc = [0x0f, 0x0e, 0x0d, 0x0c, ...],
      payload = [0x0b, 0x0a, 0x09, 0x08, ...],
    } HpkeCiphertext,
    helper_encrypted_input-share = struct {
      config_id = 2,
      enc = [0x0c, 0x0d, 0x0e, 0x0f, ...],
      payload = [0x08, 0x00, 0x0a, 0x0b, ...],

>     } HpkeCiphertext,

Reviewer 4: This may need to be indented one more level

#### Example

Successful upload

~~~ http
POST /leader/tasks/8BY0RzZMzxvA46_8ymhzycOB9krN-QIGYvg_RsByGec/reports
Host: example.com

> Content-Type: application/dap-upload-req

Reviewer: The examples should include `Content-Length` since we're going to have a MUST about the requests and responses including them.

Reviewer 4: @tgeoghegan is the answer "no" here given the discussion on https://github.com/ietf-wg-ppm/draft-ietf-ppm-dap/pull/686#discussion_r2108267711?

Reviewer: Is the answer "no" to what?

Reviewer 4: sorry, I mean this example shouldn't include the Content-Length header.

corresponding media types:

- HpkeConfigList {{hpke-config}}: "application/dap-hpke-config-list"
- UploadRequest {{upload-request}}: "application/dap-report"

> - UploadResponse {{upload-request}}: "application/dap-report-resp"

Reviewer: I think we should use `application/dap-upload-req` and `dap-upload-resp`.

| Value  | Name                     | Reference                               |
|:-------|:-------------------------|:----------------------------------------|
| `0x00` | `reserved`               | {{basic-definitions}} of RFX XXXX |
| `0x01` | `batch_collected`        | {{basic-definitions}} of RFX XXXX |
| `0x02` | `report_replayed`        | {{basic-definitions}} of RFX XXXX |
| `0x03` | `report_dropped`         | {{basic-definitions}} of RFX XXXX |
| `0x04` | `hpke_unknown_config_id` | {{basic-definitions}} of RFX XXXX |
| `0x05` | `hpke_decrypt_error`     | {{basic-definitions}} of RFX XXXX |
| `0x06` | `vdaf_prep_error`        | {{basic-definitions}} of RFX XXXX |
| `0x07` | `task_expired`           | {{basic-definitions}} of RFX XXXX |
| `0x08` | `invalid_message`        | {{basic-definitions}} of RFX XXXX |
| `0x09` | `report_too_early`       | {{basic-definitions}} of RFX XXXX |

> | `0xA`  | `task_not_started`       | {{basic-definitions}} of RFX XXXX |
> | `0xB`  | `outdated_config`        | {{basic-definitions}} of RFX XXXX |

Reviewer 4: ```suggestion
| `0x0A`  | `task_not_started`       | {{basic-definitions}} of RFX XXXX |
| `0x0B`  | `outdated_config`        | {{basic-definitions}} of RFX XXXX |
```
